networkx/algorithms/similarity.py

```diff
@@ -757,7 +757,10 @@ def match_edges(u, v, pending_g, pending_h, Ce, matched_uv=None):
         # only attempt to match edges after one node match has been made
         # this will stop self-edges on the first node being automatically deleted
         # even when a substitution is the better option
-        if matched_uv is None or len(matched_uv) == 0:
+
+        substitution_possible = len(pending_g) != 0 and len(pending_h) != 0
+        at_least_one_node_match = matched_uv is None or len(matched_uv) == 0
+        if at_least_one_node_match and substitution_possible:
```

Comment:
At face value this makes sense to me, but I'm no expert. My understanding is that this change essentially defers the edge-checking in L787 until after the self loop has been properly accounted for. The full chain of events that causes this state is less clear to me, so take my assessment with a grain of salt!

It may be worth updating the comment above this line to better explain the multiple conditions and how we think they work.